onnx/shape_inference/shape_inference.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include "onnx/proto_utils.h"
+#include "onnx/defs/schema.h"
+
+namespace ONNX_NAMESPACE { namespace shape_inference {
+
+struct InferenceContextImpl : public InferenceContext {
+  virtual const AttributeProto* getAttribute(std::string name) {
+    auto attr = attributesByName_.find(name);
+    if (attr == attributesByName_.end()) {
+      return nullptr;
+    } else {
+      return &attr->second;
+    }
+  }
+  virtual const std::vector<TypeProto_Tensor>& getAllInputTypes() {
+    return allInputTypes_;
+  }
+  virtual Status setAllOutputTypes(std::vector<TypeProto_Tensor>&& allOutputTypes) {
+    allOutputTypes_ = allOutputTypes;
+    return Status::Success;
+  }
+  std::unordered_map<std::string, AttributeProto> attributesByName_;
+  std::vector<TypeProto_Tensor> allInputTypes_;
+  std::vector<TypeProto_Tensor> allOutputTypes_;
+};
+
+void InferShapes(ONNX_NAMESPACE::ModelProto& m) {
+  auto g = m.mutable_graph();
+
+  std::unordered_map<std::string, TypeProto_Tensor> valueTypesByName;
+  {
+    for (int i = 0; i < g->value_info_size(); i++) {
+      auto vi = g->value_info(i);
+      valueTypesByName[vi.name()] = vi.type().tensor_type();
+    }
+    for (int i = 0; i < g->input_size(); i++) {
+      auto vi = g->input(i);
+      valueTypesByName[vi.name()] = vi.type().tensor_type();
+    }
+    for (int i = 0; i < g->output_size(); i++) {
+      auto vi = g->output(i);
+      valueTypesByName[vi.name()] = vi.type().tensor_type();
+    }
+  }
+
+  for (int i = 0; i < g->node_size(); i++) {
+    auto n = g->mutable_node(i);
+    auto schema = OpSchemaRegistry::Schema(n->op_type());
+    if (schema) {
+      InferenceContextImpl ctx;
+
+      {
+        std::unordered_map<std::string, AttributeProto> attributesByName;
+        for (int i = 0; i < n->attribute_size(); i++) {
+          auto attr = n->attribute(i);
+          attributesByName[attr.name()] = attr;
+        }
+        ctx.attributesByName_ = std::move(attributesByName);
+      }
+
+      {
+        std::vector<TypeProto_Tensor> allInputTypes;
+
+        for (int i = 0; i < n->input_size(); i++) {
+          auto input = n->input(i);
+          auto iter = valueTypesByName.find(input);
+          if (iter != valueTypesByName.end()) {
+            allInputTypes.push_back(iter->second);
```

Comment:
this is wrong - you can't skip inputs - the size of allInputTypes_ should be the same as n->input_size()

Reply:
++ will look at it again